src/EditorFeatures/Core/Implementation/Diagnostics/DiagnosticsClassificationTaggerProvider.cs

```diff
@@ -57,5 +64,36 @@ internal partial class DiagnosticsClassificationTaggerProvider : AbstractDiagnos
 
         protected internal override ITagSpan<ClassificationTag> CreateTagSpan(bool isLiveUpdate, SnapshotSpan span, DiagnosticData data) =>
             new TagSpan<ClassificationTag>(span, _classificationTag);
+
+        protected internal override ImmutableArray<DiagnosticDataLocation> GetLocationsToTag(DiagnosticData diagnosticData)
+        {
+            using var locationsToTagDisposer = ArrayBuilder<DiagnosticDataLocation>.GetInstance(out var locationsToTag);
+
+            // If there are 'unnecessary' locations specified in the property bag, use those instead of the main diagnostic location.
+            if (diagnosticData.AdditionalLocations != null
+                && diagnosticData.AdditionalLocations.Count > 0
+                && diagnosticData.Properties.TryGetValue(WellKnownDiagnosticTags.Unnecessary, out var unnecessaryIndices))
+            {
+                var additionalLocations = diagnosticData.AdditionalLocations.ToImmutableArray();
+                var indices = GetLocationIndices(unnecessaryIndices);
+                locationsToTag.AddRange(indices.Select(i => additionalLocations[i]).ToImmutableArray());
+            }
+            else
+            {
+                locationsToTag.Add(diagnosticData.DataLocation);
+            }
+
+            return locationsToTag.ToImmutable();
+
+            static IEnumerable<int> GetLocationIndices(string indicesProperty)
```

Comment:
Can this throw if we receive a diagnostic which has a property with key `WellKnownDiagnosticTags.Unnecessary`, but value is not in our expected format? If so, probably worth catching any parsing related exceptions and also handling that case in the caller.

Reply:
yes it can throw.  I wasn't certain if we just wanted to let it propagate?  As we control both sides it shouldn't ever be in the wrong format afaik when actually running.  If we catch / NFW someone editing the code might not realize they broke it.  Not sure.

Reply:
> As we control both sides it shouldn't ever be in the wrong format afaik

Any third party analyzer can report a diagnostic with a property bag which has a key named Unnecessary. I think we should handle this gracefully if possible.

Reply:
Ah, right.  I guess we should NFW it since we generally shouldn't expect it.  I'll look into doing that.